### PR TITLE
nodejs: fix sandboxed build on darwin

### PR DESCRIPTION
**Pull request by tie**
## Description of changes

This change fixes sandboxed build on Darwin for Node.js.

Fixes https://github.com/NixOS/nixpkgs/issues/261820

~~Note: wait for upstream PR before merging.~~
The fix has been merged upstream, this PR backports it.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- [x] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/web/nodejs/gyp-patches.nix
Comment (marsam) [inline pkgs/development/web/nodejs/gyp-patches.nix:4]: please use the commit url, since it's more stable:
```suggestion
  url = "https://github.com/nodejs/gyp-next/commit/43509c3d7d805baa2fb393b0420297d34fe398de.patch";
```
Besides that, LGTM

Comment (tie) [reply pkgs/development/web/nodejs/gyp-patches.nix]: Copied the patch with `file://` URL.

**Comment by tie** [comment]
OK, I think it’s been enough time and upstream Node.js maintainers seem to be ignoring the PR, unfortunately. Let’s copy the patch for now so that sandboxed build on macOS is fixed for 24.05 release.

File: pkgs/development/web/nodejs/darwin-sandbox.patch
Comment (wegank) [inline pkgs/development/web/nodejs/darwin-sandbox.patch:50]: Looks like doing `patchShebangs` could also work?

File: pkgs/development/web/nodejs/gyp-patches.nix
Comment (wegank) [reply pkgs/development/web/nodejs/gyp-patches.nix]: > I don’t think GitHub would keep this commit for a long time if I force push to the Node.js PR branch (e.g. to rebase it), so I’d rather wait for PR to be merged (so we can use PR patch URL), or bundle the patch in Nixpkgs PR.

It normally does. In case of doubt, we can use a link from web.archive.org. It's just too much to carry a 207-line patch that might subject to upstream changes in Nixpkgs.

File: pkgs/development/web/nodejs/darwin-sandbox.patch
Comment (tie) [reply pkgs/development/web/nodejs/darwin-sandbox.patch:50]: Yes, but then Python would be in Node.js closure for no reason.
Relevant file in the output: `lib/node_modules/npm/node_modules/node-gyp/gyp/pylib/gyp/mac_tool.py`.

We can remove references or replace with original shebang, but IMO that would be a workaround while the patch fixes the problem at its core.

File: pkgs/development/web/nodejs/gyp-patches.nix
Comment (tie) [reply pkgs/development/web/nodejs/gyp-patches.nix]: >It normally does.

I’m not so sure about this. I couldn’t find any official docs except for https://github.com/isaacs/github/issues/997, and I think I’ve seen some commits (orphaned/unreachable/overwritten by force-push) from old PRs vanish, but I might be mistaken.

Comment (tie) [reply pkgs/development/web/nodejs/gyp-patches.nix]: Using web.archive.org seems fine though. I’ll change it back to the URL.

**Review by wegank — DISMISSED** [review]

**Comment by Aleksanaa** [comment]
@ofborg build nodejs

**Review by aduh95 — APPROVED** [review]

**Comment by tie** [comment]
Reposting email I’ve sent directly to participants of https://github.com/nodejs/gyp-next/pull/216 because someone (I assume accidentally) locked the ability to comment.

cc @aduh95 @cclauss

Sorry for pinging here again, most people don’t read emails and I want to keep the discussion public.

---

Hi! Sorry for writing directly, I can’t comment on https://github.com/nodejs/gyp-next/pull/216 because someone limited the conversation on the repository 🫤

>An owner of this repository has limited the ability to comment to users that have contributed to this repository in the past.

Anyway, w.r.t.

>I hesitate because the assumption is correct for the vast majority of our Mac users and the others can use the --python option. […]

I don’t think node-gyp currently rewrites the shebang line when generating tools (see permalink below), so it’d still be using `#!/usr/bin/env python3` instead of the Python interpreter set via `--python` option.

https://github.com/nodejs/gyp-next/blob/4a8e328641dac7d6416d5061345fded81ba6fa1f/pylib/gyp/common.py#L529-L548

On that note, would it make sense to change `source[0]` to `#!{sys.executable}` instead of formatting tools command invocations with sys.executable? The only downside of I can think in this case is that POSIX shells have weird shebang handling when an executable is a text file and there is something wrong with the interpreter (see https://pubs.opengroup.org/onlinepubs/9699919799/utilities/V3_chap02.html#tag_18_09_01_01) and I usually prefer to avoid shebangs if possible for this reason (though that *shouldn’t* be an issue in this case, and we already use shebangs).
That could make the change in the PR smaller and hopefully easier to maintain. What do you think?

—
Ivan

**Comment by tie** [comment]
TODO: `test-ci-js` reaches open files limit on macOS for some reason, seems to be related to fs watch tests.

<details>
<summary>Last lines</summary>

```
not ok 3911 parallel/test-watch-mode-files_watcher
  ---
  duration_ms: 360085.73700
  severity: fail
  exitcode: -15
  stack: |-
    timeout
    TAP version 13
    # Subtest: watch mode file watcher
        # Subtest: should watch changed files
        not ok 1 - should watch changed files
          ---
          duration_ms: 3.795208
          location: 'file:///private/tmp/nix-build-nodejs-20.12.2.drv-1/node-v20.12.2/test/parallel/test-watch-mode-files_watcher.mjs:46:3'
          failureType: 'uncaughtException'
          error: 'EMFILE: too many open files, watch'
          code: 'EMFILE'
          stack: |-
            FSWatcher._handle.onchange (node:internal/fs/watchers:207:21)
          ...
        # Subtest: should debounce changes
        not ok 2 - should debounce changes
          ---
          duration_ms: 2.563541
          location: 'file:///private/tmp/nix-build-nodejs-20.12.2.drv-1/node-v20.12.2/test/parallel/test-watch-mode-files_watcher.mjs:54:3'
          failureType: 'uncaughtException'
          error: 'EMFILE: too many open files, watch'
          code: 'EMFILE'
          stack: |-
            FSWatcher._handle.onchange (node:internal/fs/watchers:207:21)
          ...
        # Subtest: should ignore files in watched directory if they are not filtered
        not ok 3 - should ignore files in watched directory if they are not filtered
          ---
          duration_ms: 1.228875
          location: 'file:///private/tmp/nix-build-nodejs-20.12.2.drv-1/node-v20.12.2/test/parallel/test-watch-mode-files_watcher.mjs:73:3'
          failureType: 'uncaughtException'
          error: 'EMFILE: too many open files, watch'
          code: 'EMFILE'
          stack: |-
            FSWatcher._handle.onchange (node:internal/fs/watchers:207:21)
          ...
        # Subtest: should allow clearing filters
        not ok 4 - should allow clearing filters
          ---
          duration_ms: 1.007208
          location: 'file:///private/tmp/nix-build-nodejs-20.12.2.drv-1/node-v20.12.2/test/parallel/test-watch-mode-files_watcher.mjs:82:3'
          failureType: 'uncaughtException'
          error: 'EMFILE: too many open files, watch'
          code: 'EMFILE'
          stack: |-
            FSWatcher._handle.onchange (node:internal/fs/watchers:207:21)
          ...
        # Subtest: should watch all files in watched path when in "all" mode
        not ok 5 - should watch all files in watched path when in "all" mode
          ---
          duration_ms: 0.661666
          location: 'file:///private/tmp/nix-build-nodejs-20.12.2.drv-1/node-v20.12.2/test/parallel/test-watch-mode-files_watcher.mjs:98:3'
          failureType: 'uncaughtException'
          error: 'EMFILE: too many open files, watch'
          code: 'EMFILE'
          stack: |-
            FSWatcher._handle.onchange (node:internal/fs/watchers:207:21)
          ...
        # Subtest: should ruse existing watcher if it exists
        ok 6 - should ruse existing watcher if it exists
          ---
          duration_ms: 0.569667
          ...
        # Subtest: should ruse existing watcher of a parent directory
        ok 7 - should ruse existing watcher of a parent directory
          ---
          duration_ms: 0.205458
          ...
        # Subtest: should remove existing watcher if adding a parent directory watcher
        ok 8 - should remove existing watcher if adding a parent directory watcher
          ---
          duration_ms: 0.567292
          ...
        # Subtest: should clear all watchers when calling clear
        ok 9 - should clear all watchers when calling clear
          ---
          duration_ms: 0.549
          ...
        # Subtest: should watch files from subprocess IPC events
        not ok 10 - should watch files from subprocess IPC events
          ---
          duration_ms: 36.674916
          location: 'file:///private/tmp/nix-build-nodejs-20.12.2.drv-1/node-v20.12.2/test/parallel/test-watch-mode-files_watcher.mjs:151:3'
          failureType: 'uncaughtException'
          error: 'EMFILE: too many open files, watch'
          code: 'EMFILE'
          stack: |-
            FSWatcher._handle.onchange (node:internal/fs/watchers:207:21)
          ...
  ...

Failed tests:
out/Release/node /private/tmp/nix-build-nodejs-20.12.2.drv-1/node-v20.12.2/test/parallel/test-fs-promises-watch.js
out/Release/node /private/tmp/nix-build-nodejs-20.12.2.drv-1/node-v20.12.2/test/parallel/test-fs-watch-encoding.js
out/Release/node /private/tmp/nix-build-nodejs-20.12.2.drv-1/node-v20.12.2/test/parallel/test-fs-watch.js
out/Release/node /private/tmp/nix-build-nodejs-20.12.2.drv-1/node-v20.12.2/test/parallel/test-fs-watch-recursive-add-file.js
out/Release/node /private/tmp/nix-build-nodejs-20.12.2.drv-1/node-v20.12.2/test/parallel/test-fs-watch-recursive-add-file-to-new-folder.js
out/Release/node /private/tmp/nix-build-nodejs-20.12.2.drv-1/node-v20.12.2/test/parallel/test-fs-watch-recursive-add-file-with-url.js
out/Release/node /private/tmp/nix-build-nodejs-20.12.2.drv-1/node-v20.12.2/test/parallel/test-fs-watch-recursive-add-folder.js
out/Release/node /private/tmp/nix-build-nodejs-20.12.2.drv-1/node-v20.12.2/test/parallel/test-fs-watch-recursive-assert-leaks.js
out/Release/node /private/tmp/nix-build-nodejs-20.12.2.drv-1/node-v20.12.2/test/parallel/test-fs-watch-recursive-add-file-to-existing-subfolder.js
out/Release/node /private/tmp/nix-build-nodejs-20.12.2.drv-1/node-v20.12.2/test/parallel/test-fs-watch-recursive-promise.js
out/Release/node /private/tmp/nix-build-nodejs-20.12.2.drv-1/node-v20.12.2/test/parallel/test-fs-watch-recursive-symlink.js
out/Release/node /private/tmp/nix-build-nodejs-20.12.2.drv-1/node-v20.12.2/test/parallel/test-fs-watch-recursive-sync-write.js
out/Release/node /private/tmp/nix-build-nodejs-20.12.2.drv-1/node-v20.12.2/test/parallel/test-fs-watch-recursive-update-file.js
out/Release/node /private/tmp/nix-build-nodejs-20.12.2.drv-1/node-v20.12.2/test/parallel/test-fs-watchfile.js
out/Release/node /private/tmp/nix-build-nodejs-20.12.2.drv-1/node-v20.12.2/test/parallel/test-macos-app-sandbox.js
out/Release/node /private/tmp/nix-build-nodejs-20.12.2.drv-1/node-v20.12.2/test/parallel/test-os.js
out/Release/node /private/tmp/nix-build-nodejs-20.12.2.drv-1/node-v20.12.2/test/parallel/test-os-process-priority.js
out/Release/node --expose-internals /private/tmp/nix-build-nodejs-20.12.2.drv-1/node-v20.12.2/test/parallel/test-runner-watch-mode.mjs
out/Release/node /private/tmp/nix-build-nodejs-20.12.2.drv-1/node-v20.12.2/test/parallel/test-runner-run.mjs
out/Release/node /private/tmp/nix-build-nodejs-20.12.2.drv-1/node-v20.12.2/test/pummel/test-fs-watch-non-recursive.js
out/Release/node /private/tmp/nix-build-nodejs-20.12.2.drv-1/node-v20.12.2/test/sequential/test-fs-watch.js
out/Release/node --expose-internals /private/tmp/nix-build-nodejs-20.12.2.drv-1/node-v20.12.2/test/parallel/test-watch-mode-files_watcher.mjs
make: *** [Makefile:550: test-ci-js] Error 1
```
</details>

[Full log download](https://github.com/user-attachments/files/15536919/log.txt)

```
$ sysctl kern.maxfiles kern.maxfilesperproc
kern.maxfiles: 122880
kern.maxfilesperproc: 61440
```

cc @aduh95 (#313982)

**Comment by n8henrie** [comment]
@tie I vaguely remember trying to wrap my head around all the fd limits on MacOS (and how they seem independently configurable but not all of them seem to take effect). Perhaps `launchctl limit maxfiles` will elucidate the issue? https://superuser.com/questions/1369590/launchctl-limit-maxfiles-vs-sysctl-kern-maxfilesperproc

File: pkgs/development/web/nodejs/gyp-patches.nix
Comment (aduh95) [inline pkgs/development/web/nodejs/gyp-patches.nix:4]:
```suggestion
  url = "https://github.com/nodejs/gyp-next/commit/706d04aba5bd18f311dc56f84720e99f64c73466.patch";
```

**Comment by cclauss** [comment]
Please resolve git conflicts.

File: pkgs/development/web/nodejs/gyp-patches.nix
Comment (tie) [reply pkgs/development/web/nodejs/gyp-patches.nix]: Changed to the [PR patch URL](https://github.com/nodejs/gyp-next/pull/216.patch) that AFAIK has stable content after merge. Also updated to use `fetchpatch2`.

File: pkgs/development/web/nodejs/v22.nix
Comment (tie) [inline pkgs/development/web/nodejs/v22.nix:19]: https://github.com/nodejs/node/commit/33242ff042200f653568a496518f9b78124ec2da is in 22.3.0

File: pkgs/development/web/nodejs/gyp-patches.nix
Comment (aduh95) [reply pkgs/development/web/nodejs/gyp-patches.nix]: Let's use the commit URL instead, the PR might get deleted or whatnot, and the actual commit that landed is what we should be fetching anyway.

File: pkgs/development/web/nodejs/v22.nix
Comment (aduh95) [inline pkgs/development/web/nodejs/v22.nix:22]:
```suggestion
      url = "https://github.com/nodejs/node/commit/14863e80584e579fd48c55f6373878c821c7ff7e.patch";
```

Comment (tie) [reply pkgs/development/web/nodejs/v22.nix:22]: Why use commit URL instead of PR patch URL?

Comment (aduh95) [reply pkgs/development/web/nodejs/v22.nix:22]: The PR might get deleted, the commit not. Also, we should pick the commit that landed on nodejs/node, not the commits that led to that commit.

Comment (tie) [reply pkgs/development/web/nodejs/v22.nix:22]: I don’t think it’s possible to delete PR without deleting repository (and the commit can be GCed if it is unreachable), but I understand your point w.r.t using landed commits. I’ll update this (and other PR URLs) tomorrow to use commit patch URLs.

File: pkgs/development/web/nodejs/nodejs.nix
Comment (aduh95) [inline pkgs/development/web/nodejs/nodejs.nix:145]: It seems to be Darwin-specific, no? If so, could we skip it only for 18.x on Darwin?

Comment (tie) [reply pkgs/development/web/nodejs/nodejs.nix]: Seems to build fine with checks on Linux, yes. Changed that to check that we are not building v18 on Darwin.

**Review by aduh95 — APPROVED** [review]

**Comment by tie** [comment]
Currently building `nodejs_{18,20,22}` on aarch64-darwin to check that everything is fine after recent rebase. Linux builds are fine.

**Comment by aduh95** [comment]
```
Cannot nix-instantiate `nodejs.passthru.tests' because:
error: attribute 'tests' in selection path 'nodejs.passthru.tests' not found
Cannot nix-instantiate `nodejs_18.passthru.tests' because:
error: attribute 'tests' in selection path 'nodejs_18.passthru.tests' not found
Cannot nix-instantiate `nodejs_22.passthru.tests' because:
error: attribute 'tests' in selection path 'nodejs_22.passthru.tests' not found
```

**Comment by tie** [comment]
We didn’t have `nodejs.passthru.tests` before, added `testVersion` in the latest commit 😅

**Comment by tie** [comment]
Successfully built `nodejs_{18,20,22}` on `aarch64-darwin`.

**Comment by tie** [comment]
Removed backport label since IIRC backports fail anyway (see https://github.com/NixOS/nixpkgs/pull/321292#issuecomment-2196552808) and it probably makes more sense to backport _all_ merged changes to stable release.

**Comment by K900** [comment]
Can we remove the Python 3.12 stuff from this PR? It's hard to tell what changes actually affect what as a mostly-unaware observer, and we've already pinned it to 3.11, which is probably fine for now.

**Comment by tie** [comment]
OK, I can split these commits into another PR.

**Comment by aduh95** [comment]
```
not ok 3036 parallel/test-tls-env-bad-extra-ca
  ---
  duration_ms: 84.14600
  severity: fail
  exitcode: 1
  stack: |-
    node:events:497
          throw er; // Unhandled 'error' event
          ^
    
    Error: spawn /build/node-v20.15.0/out/Release/node (deleted) ENOENT
        at ChildProcess._handle.onexit (node:internal/child_process:286:19)
        at onErrorNT (node:internal/child_process:484:16)
        at process.processTicksAndRejections (node:internal/process/task_queues:82:21)
    Emitted 'error' event on ChildProcess instance at:
        at ChildProcess._handle.onexit (node:internal/child_process:292:12)
        at onErrorNT (node:internal/child_process:484:16)
        at process.processTicksAndRejections (node:internal/process/task_queues:82:21) {
      errno: -2,
      code: 'ENOENT',
      syscall: 'spawn /build/node-v20.15.0/out/Release/node (deleted)',
      path: '/build/node-v20.15.0/out/Release/node (deleted)',
      spawnargs: [ '/build/node-v20.15.0/test/parallel/test-tls-env-bad-extra-ca.js' ]
    }
    
    Node.js v20.15.0
…
not ok 3043 parallel/test-tls-enable-trace
  ---
  duration_ms: 87.46300
  severity: fail
  exitcode: 1
  stack: |-
    (node:129709) internal/test/binding: These APIs are for internal testing only. Do not use them.
    (Use `node --trace-warnings ...` to show where the warning was created)
    node:events:497
          throw er; // Unhandled 'error' event
          ^
    
    Error: spawn /build/node-v20.15.0/out/Release/node (deleted) ENOENT
        at ChildProcess._handle.onexit (node:internal/child_process:286:19)
        at onErrorNT (node:internal/child_process:484:16)
        at process.processTicksAndRejections (node:internal/process/task_queues:82:21)
    Emitted 'error' event on ChildProcess instance at:
        at ChildProcess._handle.onexit (node:internal/child_process:292:12)
        at onErrorNT (node:internal/child_process:484:16)
        at process.processTicksAndRejections (node:internal/process/task_queues:82:21) {
      errno: -2,
      code: 'ENOENT',
      syscall: 'spawn /build/node-v20.15.0/out/Release/node (deleted)',
      path: '/build/node-v20.15.0/out/Release/node (deleted)',
      spawnargs: [
        '--expose-internals',
        '/build/node-v20.15.0/test/parallel/test-tls-enable-trace.js',
        'test'
      ]
    }
    
    Node.js v20.15.0
…
not ok 3070 parallel/test-runner-cli
  ---
  duration_ms: 2773.80000
  severity: fail
  exitcode: 1
  stack: |-
    node:assert:126
      throw new AssertionError(obj);
      ^
    
    AssertionError [ERR_ASSERTION]: Expected values to be strictly equal:
    
    1 !== 0
    
        at Object.<anonymous> (/build/node-v20.15.0/test/parallel/test-runner-cli.js:318:10)
        at Module._compile (node:internal/modules/cjs/loader:1358:14)
        at Module._extensions..js (node:internal/modules/cjs/loader:1416:10)
        at Module.load (node:internal/modules/cjs/loader:1208:32)
        at Module._load (node:internal/modules/cjs/loader:1024:12)
        at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:174:12)
        at node:internal/main/run_main_module:28:49 {
      generatedMessage: true,
      code: 'ERR_ASSERTION',
      actual: 1,
      expected: 0,
      operator: 'strictEqual'
    }
    
    Node.js v20.15.0
  ...
not ok 3071 parallel/test-runner-reporters
  ---
  duration_ms: 2621.54600
  severity: fail
  exitcode: 1
  stack: |-
        …
        not ok 14 - should support a custom ESM reporter from node_modules
          ---
          duration_ms: 20.931876
          location: '/build/node-v20.15.0/test/parallel/test-runner-reporters.js:115:3'
          failureType: 'testCodeFailure'
          error: "Cannot read properties of null (reading 'toString')"
          code: 'ERR_TEST_FAILURE'
          name: 'TypeError'
          stack: |-
            TestContext.<anonymous> (/build/node-v20.15.0/test/parallel/test-runner-reporters.js:119:37)
            Test.runInAsyncScope (node:async_hooks:206:9)
            Test.run (node:internal/test_runner/test:764:25)
            Test.start (node:internal/test_runner/test:668:17)
            node:internal/test_runner/test:1104:71
            node:internal/per_context/primordials:487:82
            new Promise (<anonymous>)
            new SafePromise (node:internal/per_context/primordials:455:29)
            node:internal/per_context/primordials:487:9
            Array.map (<anonymous>)
          ...
        # Subtest: should throw when reporter setup throws asynchronously
        not ok 15 - should throw when reporter setup throws asynchronously
          ---
          duration_ms: 18.085905
          location: '/build/node-v20.15.0/test/parallel/test-runner-reporters.js:126:3'
          failureType: 'testCodeFailure'
          error: |-
            Expected values to be strictly equal:
            
            null !== 7
            
          code: 'ERR_ASSERTION'
          name: 'AssertionError'
          expected: 7
          actual: ~
          operator: 'strictEqual'
          stack: |-
            TestContext.<anonymous> (/build/node-v20.15.0/test/parallel/test-runner-reporters.js:132:12)
            Test.runInAsyncScope (node:async_hooks:206:9)
            Test.run (node:internal/test_runner/test:764:25)
            Test.start (node:internal/test_runner/test:668:17)
            node:internal/test_runner/test:1104:71
            node:internal/per_context/primordials:487:82
            new Promise (<anonymous>)
            new SafePromise (node:internal/per_context/primordials:455:29)
            node:internal/per_context/primordials:487:9
            Array.map (<anonymous>)
          ...
        # Subtest: should throw when reporter errors
        not ok 16 - should throw when reporter errors
          ---
          duration_ms: 13.274049
          location: '/build/node-v20.15.0/test/parallel/test-runner-reporters.js:138:3'
          failureType: 'testCodeFailure'
          error: |-
            Expected values to be strictly equal:
            
            null !== 7
            
          code: 'ERR_ASSERTION'
          name: 'AssertionError'
          expected: 7
          actual: ~
          operator: 'strictEqual'
          stack: |-
            TestContext.<anonymous> (/build/node-v20.15.0/test/parallel/test-runner-reporters.js:142:12)
            Test.runInAsyncScope (node:async_hooks:206:9)
            Test.run (node:internal/test_runner/test:764:25)
            Test.start (node:internal/test_runner/test:668:17)
            node:internal/test_runner/test:1104:71
            node:internal/per_context/primordials:487:82
            new Promise (<anonymous>)
            new SafePromise (node:internal/per_context/primordials:455:29)
            node:internal/per_context/primordials:487:9
            Array.map (<anonymous>)
          ...
        # Subtest: should throw when reporter errors asynchronously
        not ok 17 - should throw when reporter errors asynchronously
          ---
          duration_ms: 10.379198
          location: '/build/node-v20.15.0/test/parallel/test-runner-reporters.js:148:3'
          failureType: 'testCodeFailure'
          error: |-
            Expected values to be strictly equal:
            
            null !== 7
            
          code: 'ERR_ASSERTION'
          name: 'AssertionError'
          expected: 7
          actual: ~
          operator: 'strictEqual'
          stack: |-
            TestContext.<anonymous> (/build/node-v20.15.0/test/parallel/test-runner-reporters.js:153:12)
            Test.runInAsyncScope (node:async_hooks:206:9)
            Test.run (node:internal/test_runner/test:764:25)
            Test.start (node:internal/test_runner/test:668:17)
            node:internal/test_runner/test:1104:71
            node:internal/per_context/primordials:487:82
            new Promise (<anonymous>)
            new SafePromise (node:internal/per_context/primordials:455:29)
            node:internal/per_context/primordials:487:9
            Array.map (<anonymous>)
          ...
        # Subtest: should support stdout as a destination with spec reporter
        not ok 18 - should support stdout as a destination with spec reporter
          ---
          duration_ms: 7.689268
          location: '/build/node-v20.15.0/test/parallel/test-runner-reporters.js:159:3'
          failureType: 'testCodeFailure'
          error: "Cannot read properties of null (reading 'toString')"
          code: 'ERR_TEST_FAILURE'
          name: 'TypeError'
          stack: |-
            TestContext.<anonymous> (/build/node-v20.15.0/test/parallel/test-runner-reporters.js:164:37)
            Test.runInAsyncScope (node:async_hooks:206:9)
            Test.run (node:internal/test_runner/test:764:25)
            Test.start (node:internal/test_runner/test:668:17)
            node:internal/test_runner/test:1104:71
            node:internal/per_context/primordials:487:82
            new Promise (<anonymous>)
            new SafePromise (node:internal/per_context/primordials:455:29)
            node:internal/per_context/primordials:487:9
            Array.map (<anonymous>)
          ...
        1..18
    not ok 1 - node:test reporters
      ---
      duration_ms: 2271.202088
      type: 'suite'
      location: '/build/node-v20.15.0/test/parallel/test-runner-reporters.js:15:1'
      failureType: 'subtestsFailed'
      error: '5 subtests failed'
      code: 'ERR_TEST_FAILURE'
      ...
    1..1
    # tests 18
    # suites 1
    # pass 13
    # fail 5
    # cancelled 0
    # skipped 0
    # todo 0
    # duration_ms 2312.85008
  ...
```

**Comment by tie** [comment]
Seems to be some issue on staging branch, I’ll rebase again tomorrow.

**Comment by tie** [comment]
Looks like ofborg-eval is still broken in staging (well, at least it was when I’ve started rebuilding this PR). Manual build succeeds.

---

Manual `nix build --L -f . nodejs_18 nodejs_20 nodejs_22` succeeds on aarch64-darwin.

---

```
nixpkgs-review pr \
  --checkout commit \
  --package nodejs_18 \
  --package nodejs_20 \
  --package nodejs_22 \
  262124
```

Result of `nixpkgs-review pr 262124` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>6 packages built:</summary>
  <ul>
    <li>nodejs_18</li>
    <li>nodejs_18.libv8 (nodejs_18.libv8.libv8)</li>
    <li>nodejs_20</li>
    <li>nodejs_20.libv8 (nodejs_20.libv8.libv8)</li>
    <li>nodejs_22</li>
    <li>nodejs_22.libv8 (nodejs_22.libv8.libv8)</li>
  </ul>
</details>

Result of `nixpkgs-review pr 262124` run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>6 packages built:</summary>
  <ul>
    <li>nodejs_18</li>
    <li>nodejs_18.libv8 (nodejs_18.libv8.libv8)</li>
    <li>nodejs_20</li>
    <li>nodejs_20.libv8 (nodejs_20.libv8.libv8)</li>
    <li>nodejs_22</li>
    <li>nodejs_22.libv8 (nodejs_22.libv8.libv8)</li>
  </ul>
</details>

File: pkgs/development/web/nodejs/nodejs.nix
Comment (aduh95) [inline pkgs/development/web/nodejs/nodejs.nix:149]: We want to run tests on Linux for v18.x, right?
```suggestion
    ] ++ lib.optionals (!stdenv.buildPlatform.isDarwin || lib.versionAtLeast version "20") [
```

File: pkgs/development/web/nodejs/nodejs.nix
Comment (aduh95) [inline pkgs/development/web/nodejs/nodejs.nix:161]: I think it should match the check above
```suggestion
    ] ++ lib.optionals (!stdenv.buildPlatform.isDarwin || lib.versionAtLeast version "20") [
```

Comment (tie) [reply pkgs/development/web/nodejs/nodejs.nix]: Done.

File: pkgs/development/web/nodejs/nodejs.nix
Comment (tie) [reply pkgs/development/web/nodejs/nodejs.nix]: Oops 😅
Fixed.

**Comment by tie** [comment]
Rebased, ofborg is happy now except for LLVM build timeouts on Darwin, but it builds for me on aarch64-darwin, aarch64-linux and x86_64-linux. I don’t have x86_64-darwin to test though.

**Review by aduh95 — APPROVED** [review]
It'd be nice to have the CI show it's working. Regardless of CI, we're certainly better with this patch than without it, so LGTM.

**Review by wegank — APPROVED** [review]

**Comment by aduh95** [comment]
We're unlikely to get a review from marsam: https://github.com/NixOS/nixpkgs/pull/306702

**Comment by wegank** [comment]
Oh, I forgot that, sorry...

**Pull request merged**
